python/tests/expression/test_literals.py

```diff
@@ -0,0 +1,479 @@
+
+    assert lit.value.as_tuple() == lit.to(DecimalType(9, 2)).value.as_tuple()
+    assert lit.value.as_tuple() == lit.to(DecimalType(11, 2)).value.as_tuple()
+    assert lit.to(DecimalType(9, 0)) is None
+    assert lit.to(DecimalType(9, 1)) is None
+    assert lit.to(DecimalType(9, 3)) is None
+
+
+# STRING
+
+
+def test_string_to_string_literal():
+    assert iceberg.expression.literals.of("abc") == iceberg.expression.literals.of("abc").to(StringType())
+
+
+def test_string_to_date_literal():
+    date_str = iceberg.expression.literals.of("2017-08-18")
+    date = date_str.to(DateType())
+
+    avro_val = LOGICAL_WRITERS["int-date"](datetime.strptime("2017-08-18", "%Y-%m-%d"), None)
+    assert avro_val == date.value
+
+
+def test_string_to_time_literal():
+    time_str = iceberg.expression.literals.of("14:21:01.919")
+    time_lit = time_str.to(TimeType())
+
+    avro_val = LOGICAL_WRITERS["long-time-micros"](datetime.strptime("14:21:01.919", "%H:%M:%S.%f").time(), None)
+
+    assert avro_val == time_lit.value
+
+
+def test_string_to_timestamp_literal():
+    timestamp_str = iceberg.expression.literals.of("2017-08-18T14:21:01.919+00:00")
+    timestamp = timestamp_str.to(TimestamptzType())
+
+    avro_val = LOGICAL_WRITERS["long-timestamp-micros"](dateutil.parser.parse("2017-08-18T14:21:01.919+00:00"), None)
+    assert avro_val == timestamp.value
+
+    timestamp_str = iceberg.expression.literals.of("2017-08-18T14:21:01.919")
+    timestamp = timestamp_str.to(TimestampType())
+    assert avro_val == timestamp.value
+
+    timestamp_str = iceberg.expression.literals.of("2017-08-18T14:21:01.919-07:00")
+    timestamp = timestamp_str.to(TimestamptzType())
+    avro_val = LOGICAL_WRITERS["long-timestamp-micros"](dateutil.parser.parse("2017-08-18T21:21:01.919+00:00"), None)
+    assert avro_val == timestamp.value
+
+
+def test_timestamp_with_zone_without_zone_in_literal():
+    with raises(RuntimeError):
+        timestamp_str = iceberg.expression.literals.of("2017-08-18T14:21:01.919")
+        timestamp_str.to(TimestamptzType())
+
+
+def test_timestamp_without_zone_with_zone_in_literal():
+    with raises(RuntimeError):
+        timestamp_str = iceberg.expression.literals.of("2017-08-18T14:21:01.919+07:00")
+        timestamp_str.to(TimestampType())
+
+
+def test_string_to_uuid_literal():
+    expected = uuid.uuid4()
+    uuid_str = iceberg.expression.literals.of(str(expected))
+    uuid_lit = uuid_str.to(UUIDType())
+
+    assert expected == uuid_lit.value
+
+
+def test_string_to_decimal_literal():
+    decimal_str = iceberg.expression.literals.of("34.560")
+    decimal_lit = decimal_str.to(DecimalType(9, 3))
+
+    assert 3 == abs(decimal_lit.value.as_tuple().exponent)
+    assert Decimal("34.560").as_tuple() == decimal_lit.value.as_tuple()
+
+    assert decimal_str.to(DecimalType(9, 2)) is None
+    assert decimal_str.to(DecimalType(9, 4)) is None
+
+
+# MISC
+
+
+def test_identity_conversions():
```

Comment:
parametrize from pytest would be better here since it will create a test for every pair. Something like:
```py
@pytest.mark.parametrize(
    "literal, primitive_type",
    [
        (iceberg.expression.literals.of(True), BooleanType()),
        (iceberg.expression.literals.of(34), IntegerType()),
        (iceberg.expression.literals.of(3400000000), LongType()),
        (iceberg.expression.literals.of(34.11), FloatType()),
        (iceberg.expression.literals.of(34.11), DoubleType()),
        (iceberg.expression.literals.of(Decimal(34.55).quantize(Decimal("0.01"))), DecimalType(9, 2)),
        (iceberg.expression.literals.of("2017-08-18"), DateType()),
        (iceberg.expression.literals.of("14:21:01.919"), TimeType()),
        (iceberg.expression.literals.of("2017-08-18T14:21:01.919"), TimestampType()),
        (iceberg.expression.literals.of("abc"), StringType()),
        (iceberg.expression.literals.of(uuid.uuid4()), UUIDType()),
        (iceberg.expression.literals.of(bytes([0x01, 0x02, 0x03])), FixedType(3)),
        (iceberg.expression.literals.of(bytearray([0x03, 0x04, 0x05, 0x06])), BinaryType()),
    ],
)
def test_identity_conversions(literal, primitive_type):
    expected = literal.to(primitive_type)
    assert expected is expected.to(primitive_type)
```

Reply:
Thank you Sam, I was also looking for any other test cases can be optimized, there's potentially some invalid type conversion but looks like the repeating pattern and we already get a helper method `assert_invalid_conversions` created for those, so I just ends up parameterize for this method you suggested.